qdev_wrappers/customised_instruments/parametric_sequencer.py

```diff
@@ -1,197 +1,459 @@
-                           set_cmd=self._set_seq_mode,
-                           get_cmd=self._get_seq_mode)
-        self.add_parameter(name='builder',
-                           set_cmd=self._set_sequence_builder,
-                           parameter_class=BuilderParameter)
-        self.add_parameter(name='inner_setpoints',
-                           initial_val=self.default_inner_setpoints,
-                           parameter_class=DictWithDefaultParameter)
-        self.add_parameter(name='outer_setpoints',
-                           initial_val=self.default_outer_setpoints,
-                           parameter_class=DictWithDefaultParameter)
-        self.add_parameter(name='default_builder_parms',
-                           initial_val=default_builder_parms,
-                           parameter_class=DictWithDefaultParameter)
-        self.add_parameter(name='sequence_element_val',
-                           set_cmd=self._set_seq_elem_val,
-                           get_cmd=self._get_seq_elem_val)
-
-    def _set_seq_elem_val(self, value):
-        """
-        assumes records == elements, and is for use when seq mode is off
-        """
-        if self.seq_mode():
-            raise RuntimeError('must turn off sequence mode to'
-                               ' set output of one element on loop')
-        array = np.asarray(self.inner_setpoints()['setpoints'])
-        idx = (np.abs(array - value)).argmin()
-        self.awg.sequence_pos(idx)
-
-    def _get_seq_elem_val(self, value):
-        """
-        assumes records == elements, and is for use when seq mode is off
-        """
-        if self.seq_mode():
-            raise RuntimeError('must turn off sequence mode to'
-                               ' set output of one element on loop')
-        array = np.asarray(self.inner_setpoints()['setpoints'])
-        idx = (np.abs(array - value)).argmin()
-        return array[self.awg.sequence_pos(idx)]
-
-    def _set_seq_mode(self, status):
-        """
-        not sure if seq mode needs to also live in the builder parms,
-        probably that can go, otherwise point is to make awg play each elem
-        infinitely many times until you manually switch to the next one
-        """
-        self.default_builder_parms({'seq_mode': status})
-        if status:
-            for i in range(self.sequence_length()):
-                self.set_sqel_loopcnt(i, 1)
-        else:
-            for i in range(self.sequence_length()):
-                self.set_sqel_loopcnt_to_inf(i)
 
-    def _get_seq_mode(self):
-        if self.set_sqel_loopcnt() == '1':
-            awg_val = False
+        # add metadata, that gets added to the snapshot automatically
+        # add it before the upload so that if there is a crash, the
+        # state that is causing the crash is captured in the metadata
+        # TODO: add serialization of the elements
+        self.metadata['template_element'] = template_element
```

Comment:
maybe leave this commented until the serialisation is done so that we can in the meantime add the instrument to the station?